File: packages/crypto/src/storage/wrap.ts
import { aeadEncrypt, aeadDecrypt, randomBytes, AEAD_NONCE_LEN } from '../primitives';
import { concatBytes } from '../encoding';

/**
 * Wraps arbitrary key material (or any sensitive bytes — a serialized ratchet
 * state, a private key) for storage in IndexedDB, under the Argon2id-derived KEK —
 * docs/05-crypto-architecture.md#local-key-storage. Uses the same ChaCha20-Poly1305
 * AEAD as message encryption (primitives.ts) rather than introducing a second
 * cipher for this — one audited AEAD construction to review, not two.
 *
 * Output layout: `nonce (12 bytes) || ciphertext+tag`. The nonce is fresh random
 * per call — wrapping is infrequent enough (once per session-state save, not
 * per-message) that a random 96-bit nonce's collision probability is negligible, so
 * there's no need for the counter-based nonce discipline the message ratchet uses.
 */
export function wrapBytes(kek: Uint8Array, plaintext: Uint8Array): Uint8Array {
  const nonce = randomBytes(AEAD_NONCE_LEN);
  const ciphertext = aeadEncrypt(kek, nonce, plaintext, new Uint8Array());
  return concatBytes(nonce, ciphertext);
}

export function unwrapBytes(kek: Uint8Array, wrapped: Uint8Array): Uint8Array {
  const nonce = wrapped.slice(0, AEAD_NONCE_LEN);
  const ciphertext = wrapped.slice(AEAD_NONCE_LEN);
  return aeadDecrypt(kek, nonce, ciphertext, new Uint8Array());
}

File: apps/mobile/lib/crypto/storage/wrap.dart
/// Direct port of `packages/crypto/src/storage/wrap.ts` — see that file's docstring
/// for the design rationale (ChaCha20-Poly1305 reused rather than a second cipher,
/// random-not-counter nonce since wrapping is infrequent). Output layout is
/// identical: `nonce (12 bytes) || ciphertext+tag`.
library;

import 'dart:typed_data';
import '../encoding.dart';
import '../primitives.dart';

Future<Uint8List> wrapBytes(Uint8List kek, Uint8List plaintext) async {
  final nonce = randomBytes(aeadNonceLen);
  final ciphertext = await aeadEncrypt(kek, nonce, plaintext, Uint8List(0));
  return concatBytes([nonce, ciphertext]);
}

Future<Uint8List> unwrapBytes(Uint8List kek, Uint8List wrapped) async {
  final nonce = wrapped.sublist(0, aeadNonceLen);
  final ciphertext = wrapped.sublist(aeadNonceLen);
  return aeadDecrypt(kek, nonce, ciphertext, Uint8List(0));
}
